sql/src/main/java/io/crate/metadata/sys/TableHealthService.java

```diff
@@ -0,0 +1,250 @@
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+@Singleton
+public class TableHealthService extends AbstractComponent {
+
+    private static final String STMT = "select table_name, schema_name, partition_ident, routing_state," +
+                                       " \"primary\", relocating_node, count(*) from sys.shards " +
+                                       " group by 1, 2, 3, 4, 5, 6";
+    private static final Statement PARSED_STMT = SqlParser.createStatement(STMT);
+
+    private final ClusterService clusterService;
+    private final Schemas schemas;
+    private final Provider<SQLOperations> sqlOperationsProvider;
+    private Session session;
+
+    @Inject
+    public TableHealthService(Settings settings,
+                              ClusterService clusterService,
+                              Schemas schemas,
+                              Provider<SQLOperations> sqlOperationsProvider) {
+        super(settings);
+        this.clusterService = clusterService;
+        this.schemas = schemas;
+        this.sqlOperationsProvider = sqlOperationsProvider;
+    }
+
+    public CompletableFuture<Iterable<TableHealth>> computeResults() {
+        if (clusterService.localNode() == null) {
+            /*
+              During a long startup (e.g. during an upgrade process) the localNode() may be null
+              and this would lead to NullPointerException in the TransportExecutor.
+             */
+            logger.debug("Could not retrieve tables health information. localNode is not fully available yet.");
+            return completedFuture(Collections.emptyList());
+        }
+
+        try {
+            CompletableFuture<Map<TablePartitionIdent, ShardsInfo>> future = new CompletableFuture<>();
+            HealthResultReceiver resultReceiver = new HealthResultReceiver(future);
+            session().quickExec(STMT, stmt -> PARSED_STMT, resultReceiver, Row.EMPTY);
+            return future.thenApply(this::buildTablesHealth);
+        } catch (Throwable t) {
+            logger.error("error retrieving tables health information", t);
+            return failedFuture(t);
+        }
+    }
+
+    private Session session() {
+        if (session == null) {
+            session = sqlOperationsProvider.get().newSystemSession();
+        }
+        return session;
+    }
+
+    private List<TableHealth> buildTablesHealth(Map<TablePartitionIdent, ShardsInfo> tables) {
+        List<TableHealth> tableHealthList = new ArrayList<>(tables.size());
+        for (Map.Entry<TablePartitionIdent, ShardsInfo> entry : tables.entrySet()) {
+            TablePartitionIdent ident = entry.getKey();
+            ShardsInfo shardsInfo = entry.getValue();
+            TableIdent tableIdent = new TableIdent(
+                BytesRefs.toString(ident.tableSchema), BytesRefs.toString(ident.tableName));
+            DocTableInfo tableInfo = schemas.getTableInfo(tableIdent);
```

Comment:
could there be a race condition where we `tables` contains entries for tables that got deleted so that this might raise an exception here?

Reply:
yes, good catch